Typography block support: add typography support and defaults

ramonjd opened the pull request:
Resolves: https://github.com/WordPress/gutenberg/issues/35604

## What this PR does

This PR is a follow-up to https://github.com/WordPress/gutenberg/pull/33744, which updates typography controls to use the very excellent ToolsPanel.

We're trying to achieve the following:
- Opt-in to typography design tools for text blocks. The tools will appear in the ToolsPanel and will be toggleable. Try saying that 10 times fast.
- Assign useful ToolsPanels "defaults", which are typography tools that will show (and always show) by default.


## Testing

Get comfortable. 

Rebase this branch on top of `origin/update/typography-support-to-use-new--panel`

Create a new post and insert the follow blocks:

- Button
- Code
- Heading
- List
- Navigation
- Paragraph
- Post Author
- Post Comments
- Post Comments Count
- Post Comments Form
- Post Comments Link
- Post Date
- Post Excerpt
- Post Categories (Post terms block)
- Post Tags  (Post terms block)
- Post Title
- Preformatted
- Query Pagination (next and previous blocks)
- Query Title
- Site Tagline
- Site Title
- Term Description
- Verse

For each block, please check that:

1. The Typography controls appear in the ToolsPanel and,
2. The properties defined under __experimentalDefaultControls in the block's block.json are displayed in the panel by default. 
3. Take a small break, think of something nice, like cheese. 

## Screenshots 

Lo! The Verse block opts in to `fontSize` only!

<img width="500" alt="Screen Shot 2021-08-13 at 3 37 41 pm" src="https://user-images.githubusercontent.com/6458278/129310284-ab136224-0003-4200-a020-c341905a14f2.png">

The paragraph opts in to both `fontSize` and `lineHeight`. WOOT!

<img width="500" alt="Screen Shot 2021-08-13 at 3 37 09 pm" src="https://user-images.githubusercontent.com/6458278/129310303-d11a7996-3cd3-4188-a426-50192e5ce679.png">

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

apeatling commented:
Looks good -- but why no line height on list and preformatted?

ramonjd commented:
Thanks for testing!
>  Looks good -- but why no line height on list and preformatted?

At the time, there wasn't a clear answer as to whether we should include line height for blocks that do not yet explicitly support in the typography settings.

I thought it safer to only add defaults for the current settings until there was a requirement or direction, or at least until we can do a spot of design testing.

 [A more recent comment](https://github.com/WordPress/gutenberg/pull/33744#issuecomment-898229300) from @jasmussen in the Typography: Switch to ToolsPanel for display UI PR seems to support that view.

Happy to give line height support a quick test for these blocks as well if we'd like to enable it immediately. I don't expect too much trouble from such a property. 

There's also the option to go with currently-defined properties as defaults and follow up after testing/consultation.

ramonjd commented:
Rebasing this PR, I noticed that the post date block has introduced `__experimentalFontWeight`.

For that block in particular, I don't think it looks too busy with the fontWeight enabled as well. It could be useful since there are no inline controls to adjust font weight for that block.

**Post date block with fontWeight controls**
<img width="786" alt="Screen Shot 2021-08-16 at 10 44 57 am" src="https://user-images.githubusercontent.com/6458278/129497995-abce4e81-d1f8-4b54-88e1-95dc87968d96.png">


**Post date block without fontWeight controls**
<img width="793" alt="Screen Shot 2021-08-16 at 10 44 42 am" src="https://user-images.githubusercontent.com/6458278/129498001-9d1474c8-a298-4044-b27b-3396e7aa2652.png">

apeatling commented:
> At the time, there wasn't a clear answer as to whether we should include line height for blocks that do not yet explicitly support in the typography settings.

I would have expected list item to support line height, but based on Joen's comment we could leave it out and come back to it.

ramonjd committed:
Initial commit. Adding typography fontSize and lineHeight supports defaults in block.json where the block supports each.

mtias commented on packages/block-library/src/heading/block.json:
```diff
@@ -35,7 +35,11 @@
+			"__experimentalFontWeight": true,
+			"__experimentalDefaultControls": {
+				"fontSize": true,
+				"lineHeight": true
```
We should add "weight" as a default and remove line-height

mtias commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -37,7 +37,11 @@
+			"lineHeight": true,
+			"__experimentalDefaultControls": {
+				"fontSize": true,
+				"lineHeight": true
```
We should remove line-height from defaults

mtias commented:
To expand on the review comments, I think `line-height` should never be a default; font-weight (and maybe appearance) should be available by default on "title" like blocks (headings, post title, site title, etc). Site title maybe should have all toggled by default.

A possible classification:

- Basic: only font-size by default (text, lists, etc).
- Titles: incorporates weight and style, _maybe_ also text-transform (heading, post title, etc).
- Special (some examples):
  - Site title: could have all typography options toggled.
  - Quote: should be “basic” but can also have “style” available, since italic quotes are fairly common.

ramonjd committed:
Removing line height typography default support
Opting-in to all typography supports for text blocks (except font-family for code block)
Site title: toggle all typography options as default
Other titles (heading, post title, query title): add weight and style, and text-transform defaults

ramonjd committed:
Using `fontAppearance` in __experimentalDefaultControls to flag that the fontStyle/fontWeight support control should show as default.

ramonjd commented:
Looks like from the E2E fixture tests that `core/quote` deprecations need an update to handle style supports. 
👀 at this now.

_Edit_: https://github.com/WordPress/gutenberg/pull/35615

ramonjd committed:
Removing Font family and Text decoration opt-in until a second pass.

Apply suggestions from code review

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

Reinstating typography opt-ins pending further advice

aaronrobertshaw commented on packages/block-library/src/button/block.json:
```diff
@@ -63,7 +63,10 @@
 		"typography": {
 			"fontSize": true,
-			"__experimentalFontFamily": true
+			"__experimentalFontFamily": true,
```
Based on this https://github.com/WordPress/gutenberg/issues/35604#issuecomment-944152637, should the support for Font Family be excluded here? Did we plan to maintain any support that was already enabled?

The font family support was added in https://github.com/WordPress/gutenberg/pull/30394.

ramonjd replied:
> Did we plan to maintain any support that was already enabled?

Good question. 

For this block, I seem to remember deciding not to exclude it on purpose since it had already been enabled. 

I don't have any background as to why we shouldn't be exposing font-family in the block editor. Asking over in https://github.com/WordPress/gutenberg/issues/35604#issuecomment-950040099

aaronrobertshaw commented on packages/block-library/src/code/block.json:
```diff
@@ -16,7 +16,16 @@
+			"__experimentalFontWeight": true,
+			"__experimentalLetterSpacing": true,
+			"__experimentalTextTransform": true,
+			"__experimentalTextDecoration": true,
```
The code block is a bit of a special case but is there a benefit to opting into text-decoration support? My thinking was it probably falls under the banner of a more specific revision as stated in https://github.com/WordPress/gutenberg/issues/35604#issuecomment-944152637
```suggestion
```

aaronrobertshaw commented on packages/block-library/src/heading/block.json:
```diff
@@ -35,7 +35,15 @@
+			"__experimentalDefaultControls": {
+				"fontSize": true,
+				"fontAppearance": true,
+				"__experimentalTextTransform": true
```
There aren't nested `__experimental` prefixes on `__experimentalDefaultControls`.  As it currently sits, the text-transform control does not show by default.

```suggestion
				"textTransform": true
```

aaronrobertshaw commented on packages/block-library/src/navigation/block.json:
```diff
@@ -94,8 +94,9 @@
 			"__experimentalFontWeight": true,
 			"__experimentalTextTransform": true,
-			"__experimentalFontFamily": true,
-			"__experimentalTextDecoration": true
```
I take it we are not looking to add any form of deprecation when we remove font-family support? 

It was recently switched to use CSS class names in https://github.com/WordPress/gutenberg/pull/31910. In my testing, the block is not invalidated switching to this branch but the previous font family class is lost and so to therefore is the user's selection.

ramonjd replied:
Is this in the same category of question as https://github.com/WordPress/gutenberg/pull/34064/files#r734321351?

That is, should we remove font-family support from blocks that already support it?

Prima facie I think "no", but I've asked here: https://github.com/WordPress/gutenberg/pull/34064#pullrequestreview-786542895

This was an oversight on my part. I will reinstate it. Thanks for spotting!

aaronrobertshaw replied:
My first thought is we would have had a reason to opt into the support initially. So, I'd be inclined not to take something away from users without a clear and strategic decision on doing so.

I think [your comment](https://github.com/WordPress/gutenberg/issues/35604#issuecomment-950040099) asking for clarity on this should have us covered.

aaronrobertshaw commented on packages/block-library/src/post-terms/block.json:
```diff
@@ -27,7 +27,10 @@
 			"lineHeight": true,
 			"fontSize": true,
-			"__experimentalFontWeight": true
+			"__experimentalFontWeight": true,
```
It feels like this block should also have the extra features enabled. Maybe this just sticks out as the first block that really restricts which supports are opted into.

```suggestion
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextTransform": true,
```

aaronrobertshaw commented on packages/block-library/src/post-title/block.json:
```diff
@@ -38,10 +38,15 @@
+			"__experimentalDefaultControls": {
+				"fontSize": true,
+				"fontAppearance": true,
+				"__experimentalTextTransform": true
```
No `__experimental` prefix is needed here. The text-transform control doesn't currently display as a default.

```suggestion
				"textTransform": true
```

aaronrobertshaw commented on packages/block-library/src/query-title/block.json:
```diff
@@ -26,7 +26,15 @@
+			"__experimentalDefaultControls": {
+				"fontSize": true,
+				"fontAppearance": true,
+				"__experimentalTextTransform": true
```
This also needs a tweak.

```suggestion
				"textTransform": true
```

aaronrobertshaw commented on packages/block-library/src/site-title/block.json:
```diff
@@ -36,11 +36,17 @@
+				"lineHeight": true,
+				"fontAppearance": true,
+				"__experimentalLetterSpacing": true,
+				"__experimentalTextTransform": true
```
```suggestion
				"letterSpacing": true,
				"textTransform": true
```

aaronrobertshaw reviewed (commented):
Thanks for the effort wrangling all these block supports and default controls @ramon 👍 

For the most part, these changes worked as expected. 

I rebased this PR and addressed the minor conflicts locally, then went through each block included, added it to a post, and cross-checked the supports and default controls.

I've added a few suggestions to fix up the default control flags to remove their `__experimental` prefix. When I added the `__experimentalDefaultControls` flag as part of the `ToolsPanel` I didn't see the benefit in nesting `__experimental` prefixes.

I also had a few small questions and added comments for those. Happy to give this another run through and test early next week.

ramonjd committed:
Reinstating typography opt-ins pending further advice

aaronrobertshaw commented:
All the default controls appear now as defined within the respective `block.json` files. 👍 

This is looking good for once we [get clarification on maintaining previous supports](https://github.com/WordPress/gutenberg/issues/35604#issuecomment-950040099) or not.

ramonjd commented:
> All the default controls appear now as defined within the respective block.json files. 👍

Thanks, I appreciate the 👀 once again 👍

ramonjd commented:
@mtias Are there any objections to getting this one in before Monday, and dealing with any supports removal and associated deprecations in follow ups?

I think the outstanding question was [whether to remove font-family](https://github.com/WordPress/gutenberg/issues/35604#issuecomment-950040099).

Also, whether the defaults and supports we've been testing in the latest version of this PR are satisfactory. 

🍺

apeatling reviewed (commented):
@ramonjd I think this is looking good, I think we should remove font family and then talk to Matías on the call tomorrow to see about following up with this in another PR? Alternatively you could wait until tomorrow and ask him.

ramonjd commented:
>  I think we should remove font family and then talk to Matías on the call tomorrow to see about following up with this in another PR? Alternatively you could wait until tomorrow and ask him.

We removed font-family in a previous commit, but it broke several blocks validations: https://github.com/WordPress/gutenberg/issues/35604#issuecomment-950040099

My proposal was to leave font-family for the blocks where font-family already exists and then, if we decide to remove it for these blocks, follow up with individual PRs for each deprecation (around eight).

francescamarano commented:
> @ramonjd I think this is looking good, I think we should remove font family and then talk to Matías on the call tomorrow to see about following up with this in another PR? Alternatively you could wait until tomorrow and ask him.

Hello 👋 Is the call with Matias a call for Gutenberg contributors? Thanks

ramonjd commented:
Ciao @francescamarano! I think it's a quick, on-the-fly sort of thing, so not a formal contributor meeting. Anything we learn will be posted right here. Thanks!

apeatling commented:
@francescamarano It's an informal meet and greet for our team. Not specific to Gutenberg or this PR. Anything related to it that comes up we'll post here in detail. 👍

andrewserong commented on packages/block-library/src/post-comments-count/block.json:
```diff
@@ -18,7 +18,14 @@
+			"__experimentalFontStyle": true,
+			"__experimentalFontWeight": true,
+			"__experimentalLetterSpacing": true,
+			"__experimentalTextTransform": true,
```
Does text transform make sense here? I _think_ it's always a number, so letter case has no effect?

ramonjd replied:
It does not make sense. Good spotting.

andrewserong commented on packages/block-library/src/post-comments-form/block.json:
```diff
@@ -19,7 +19,14 @@
-			"lineHeight": true
+			"lineHeight": true,
+			"__experimentalFontStyle": true,
+			"__experimentalFontWeight": true,
```
I found font weight and font size, etc, didn't _quite_ work as I expected with this block, because it's a more complex block. E.g. the font weight gets applied at the container, but there is a nested `h3` that doesn't wind up being affected by this:

![image](https://user-images.githubusercontent.com/14988353/140006207-822a60d6-fa2a-40eb-a137-dce3d9acb776.png)

Because this is a server-rendered block, I wouldn't consider these details a blocker — I think personally it's fine to opt-in to these Typography supports and defer fine-tuning the behaviour to follow-ups further down the track.

andrewserong reviewed (approved):
Nice work @ramonjd, thanks for working through all these! They're testing well for me. I've noted that a couple of the blocks behave in slightly awkward ways (e.g. it'd be great to have more granular control over typography within elements of the block), but I don't think any of these are blockers for getting this PR merged.

The one issue that we might want to tweak, is that the post comments count block currently opts-in to text transform, but since it always renders a number, I think that might not be necessary?

My full testing notes are below. Apologies for the big list!

✅ Button block: Font size shown by default
✅ Code block: Font size shown by default, all others working well
✅ Heading block: font size, appearance, text transform shown by default, all others working well
✅ List block: font size shown by default, all others working well
✅ Navigation block: font size shown by default, all others working well
✅ Paragraph block: font size shown by default, all others working well
✅ Post author block: font size shown by default, all others working well
✅ Post comment date block: font size shown by default, all others working well
✅ Post comments count block: font size shown by default, all others working well except text transform which might not make sense if it's always just a number? ❓
❓ Post comments form block: mostly works but the nested `h3` element doesn't get targeted by some of the Typography settings set at the container — left a note, but likely things to be followed up with separately for that block
✅ Post comments link: font size shown by default, all others working well
❓ Post comments block: mostly works but I ran into a similar issue as with the Post comments form block: it's a more complex block, so applying the Typography controls at the container level applies to all elements within it. But, since it's server-rendered, it's maybe not such a big issue to have it enabled
✅ Post date block: font size shown by default, all others working well
✅ Post excerpt block: font size shown by default, all others working well
✅ Post navigation block: font size shown by default, all others working well
✅ Post terms block: font size shown by default, all others working well
✅ Post title block: font size, appearance, letter case shown by default, all others working well
✅ Pre-formatted block: font size shown by default, all others working well
✅ Pullquote block: font size and appearance shown by default, all others working well
✅ Query pagination next block: font size shown by default, all others working well
✅ Query pagination prev block: font size shown by default, all others working well
✅ Query title block: font size, font appearance, text transform shown by default, all others working well
✅ Quote block: font size and appearance shown by default, all others working well
✅ Site tagline block: font size shown by default, all others working well
✅ Site title block: font size, line height, appearance, letter spacing, text transform shown by default, font family also works well
❓ Table block: all the opted-in settings work, but because we don't have separate typography controls for table heading versus content and footer, is it an issue that we're setting all these at the container level? I could imagine folks finding it frustrating that they can't control these things separately... but I suppose it's better to expose _some_ control than none!
✅ Term description block: Font size shown by default, line height can be accessed
✅ Verse block: Font size and appearance shown by default, all others working well

As I mentioned, I think some of these questions would be good things to explore in follow-ups over time, but not blockers for getting this larger PR merged in the shorter-term 👍

ramonjd committed:
Remove __experimentalTextTransform

since the rendered output is a number

ramonjd commented:
Thanks again for the always helpful review, and the eagle eyes.

> ❓ Table block: all the opted-in settings work, but because we don't have separate typography controls for table heading versus content and footer, is it an issue that we're setting all these at the container level? I could imagine folks finding it frustrating that they can't control these things separately... but I suppose it's better to expose some control than none!

This is what I found, and also came to the same conclusion as you: the typography controls are nice to have, but it seems like another one of those blocks that demand finer control over the inner elements.

🤷 

>  Post comments block: mostly works but I ran into a similar issue as with the Post comments form block: it's a more complex block, so applying the Typography controls at the container level applies to all elements within it. But, since it's server-rendered, it's maybe not such a big issue to have it enabled

>  Post comments form block: mostly works but the nested h3 element doesn't get targeted by some of the Typography settings set at the container — left a note, but likely things to be followed up with separately for that block

I'll work on immediate follow ups for these. 🍺

apeatling commented:
@francescamarano The main points that came up in a chat with Matías about this:

* We should keep font family as an option if it exists, but it should not be the default for any blocks
* We don't yet have a good way to add custom fonts, so it's quite limited.
* Once we have a better fonts API we can surface this more in blocks that are more likely to need font family changes (site title being an obvious one).

francescamarano commented:
> @francescamarano The main points that came up in a chat with Matías about this:
> 
> * We should keep font family as an option, but it should not be the default for any blocks
> * We don't yet have a good way to add custom fonts, so it's quite limited.
> * Once we have a better fonts API we can surface this more in blocks that are more likely to need font family changes (site title being an obvious one).

Thanks! Hopefully, the [Webfonts API](https://core.trac.wordpress.org/ticket/46370) will make it to WP 6.0 🚀
